toolchain/check/check.cpp

```diff
@@ -148,21 +155,72 @@ static auto TrackImport(
-    // in this case because both come from the same file.
-    auto packaging = unit_info.unit->parse_tree->packaging_directive();
-    if (packaging && import.package_id == packaging->names.package_id &&
-        import.library_id == packaging->names.library_id) {
+    // True if the file's package is implicitly `Main` (by omission).
+    bool is_file_implicit_main =
+        !packaging || !packaging->names.package_id.is_valid();
+    // True if the import is using "current package" syntax (by omission).
+    bool is_import_current_package = !import.package_id.is_valid();
+    // True if the import is using `default` library syntax.
+    bool is_import_default_library = !import.library_id.is_valid();
+    // True if the import and file point at the same package, even by
+    // incorrectly specifying the current package name to `import`.
+    bool is_same_package =
+        is_import_current_package || import.package_id == file_package_id;
+    // True if the import points at the same library as the file's library.
+    bool is_same_library =
+        is_same_package &&
+        (packaging ? import.library_id == packaging->names.library_id
+                   : is_import_default_library);
+
+    // Diagnose explicit imports of the same library, whether from `api` or
+    // `impl`.
+    if (is_same_library) {
       CARBON_DIAGNOSTIC(ExplicitImportApi, Error,
                         "Explicit import of `api` from `impl` file is "
                         "redundant with implicit import.");
       CARBON_DIAGNOSTIC(ImportSelf, Error, "File cannot import itself.");
-      unit_info.emitter.Emit(
-          import.node, packaging->api_or_impl == Parse::Tree::ApiOrImpl::Impl
-                           ? ExplicitImportApi
-                           : ImportSelf);
+      bool is_impl =
+          !packaging || packaging->api_or_impl == Parse::Tree::ApiOrImpl::Impl;
+      unit_info.emitter.Emit(import.node,
+                             is_impl ? ExplicitImportApi : ImportSelf);
       return;
     }
+
+    // Diagnose explicit imports of `Main//default`. There is no `api` for it.
+    // This lets other diagnostics handle explicit `Main` package naming.
+    if (is_file_implicit_main && is_import_current_package &&
+        is_import_default_library) {
+      CARBON_DIAGNOSTIC(ImportMainDefaultLibrary, Error,
+                        "Cannot import `Main//default`.");
+      unit_info.emitter.Emit(import.node, ImportMainDefaultLibrary);
+
+      return;
+    }
+
+    if (!is_import_current_package) {
+      // Diagnose explicit imports of the same package that use the package
+      // name.
+      if (is_same_package || (is_file_implicit_main && is_explicit_main)) {
+        CARBON_DIAGNOSTIC(
+            ImportCurrentPackageByName, Error,
+            "Imports from the current package must omit the package name.");
+        unit_info.emitter.Emit(import.node, ImportCurrentPackageByName);
```

Comment:
Seems like it would be a bit better to emit this on the actual package name parse node?

Reply:
At present, I'm just trying to issue diagnostics. There's a separate discussion about what level of detail to put into diagnostics, and how.